Objects/dictobject.c

```diff
@@ -7124,50 +7124,135 @@ set_dict_inline_values(PyObject *obj, PyDictObject *new_dict)
     }
 }
 
-int
-_PyObject_SetManagedDict(PyObject *obj, PyObject *new_dict)
+#ifdef Py_GIL_DISABLED
+
+// Trys and sets the dictionary for an object in the easy case when our current
+// dictionary is either completely not materialized or is a dictionary which
+// does not point at the inline values.
+static bool
+try_set_dict_inline_only_or_other_dict(PyObject *obj, PyObject *new_dict, PyDictObject **cur_dict)
+{
+    bool replaced = false;
+    Py_BEGIN_CRITICAL_SECTION(obj);
+
+    PyDictObject *dict = *cur_dict = _PyObject_GetManagedDict(obj);
+    if (dict == NULL) {
+        // We only have inline values, we can just completely replace them.
+        set_dict_inline_values(obj, (PyDictObject *)new_dict);
+        replaced = true;
+        goto exit_lock;
+    }
+
+    if (FT_ATOMIC_LOAD_PTR_RELAXED(dict->ma_values) != _PyObject_InlineValues(obj)) {
+        // We have a materialized dict which doesn't point at the inline values,
+        // We get to simply swap dictionaries and free the old dictionary.
+        FT_ATOMIC_STORE_PTR(_PyObject_ManagedDictPointer(obj)->dict,
+                (PyDictObject *)Py_XNewRef(new_dict));
+        replaced = true;
+        goto exit_lock;
+    } else {
+        // We have inline values, we need to lock the dict and the object
+        // at the same time to safely dematerialize them. To do that while releasing
+        // the object lock we need a strong reference to the current dictionary.
+        Py_INCREF(dict);
+    }
+exit_lock:
+    Py_END_CRITICAL_SECTION();
+    return replaced;
+}
+
+// Replaces a dictionary that is probably the dictionary which has been
+// materialized and points at the inline values. We could have raced
+// and replaced it with another dictionary though.
+static int
+replace_dict_probably_inline_materialized(PyObject *obj, PyDictObject *inline_dict,
+                                          PyObject *new_dict, bool clear,
+                                          PyDictObject **replaced_dict)
+{
+    // But we could have had another thread race in after we released
+    // the object lock
+    int err = 0;
+    *replaced_dict = _PyObject_GetManagedDict(obj);
+    assert(FT_ATOMIC_LOAD_PTR_RELAXED(inline_dict->ma_values) == _PyObject_InlineValues(obj));
+
+    if (*replaced_dict == inline_dict) {
+        err = _PyDict_DetachFromObject(inline_dict, obj);
+        if (err != 0) {
+            return err;
+        }
+        // We incref'd the inline dict and the object owns a ref.
+        // Clear the object's reference, we'll clear the local
+        // reference after releasing the lock.
+        if (clear) {
+            Py_XDECREF((PyObject *)*replaced_dict);
+        } else {
+            _PyObject_XDecRefDelayed((PyObject *)*replaced_dict);
+        }
+        *replaced_dict = NULL;
```

Comment:
It should be safe because we have 2 references here. The first is the reference the object owns, and the 2nd is the reference we acquire in `try_set_dict_inline_only_or_other_dict`. That 2nd reference is necessary for locking the dictionary (otherwise we can free it before the `Py_END_CRITICAL_SECTION2` in `set_or_clear_managed_dict`).

I tried to capture this in the comment in here as obviously it's tricky, but I'll expand on it to cover that the replacement coming after is still safe.

I think the main reason to not assign the dict first is just that the `_PyDict_DetachFromObject` code really wants the dictionary that's being detached to be the current dictionary, and given the dictionary remains valid (and the object is locked) the order shouldn't matter.

I'll combine the decref's, I think this also makes it more obvious that we may need to delay dec-ref'd prev_dict in the caller after calling this function as well.